File: stash/347.前-k-个高频元素.cpp
/*
 * @lc app=leetcode.cn id=347 lang=cpp
 *
 * [347] 前 K 个高频元素
 */

// @lc code=start
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> counts;
        // 计算频次
        for(int i : nums) counts[i]++;
        // 最小堆
        priority_queue<pair<int,int>, vector<pair<int,int>>, greater<pair<int,int>>> q;
        // 堆中元素为 [频次，数值] 元组，并根据频次维护小顶堆特性
        for(auto it : counts) {
            if (q.size() != k) {
                q.push(make_pair(it.second, it.first));
            } else {
                if (it.second > q.top().first) {
                    q.pop();
                    q.push(make_pair(it.second, it.first));
                }
            }
        }
        vector<int> res;
        while(q.size()) {
            res.push_back(q.top().second);
            q.pop();
        }
        return vector<int>(res.rbegin(), res.rend());
    }
};
// @lc code=end

File: stash/347.前-k-个高频元素.js
/*
 * @lc app=leetcode.cn id=347 lang=javascript
 *
 * [347] 前 K 个高频元素
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var topKFrequent = function (nums, k) {
  const counts = {};
  for (let num of nums) {
    counts[num] = (counts[num] || 0) + 1;
  }
  let pairs = Object.keys(counts).map(key => [counts[key], key]);
  console.log(pairs);
  select(0, pairs.length - 1, k);
  console.log(pairs);
  return pairs.slice(0, k).map(item => item[1]);

  function select(left, right, offset) {
    if (left <= right) {
      return;
    }
    const pivotIndex = partition(0, pairs.length - 1);
    if (pivotIndex === offset) {
      return;
    }

    console.log(pairs);

    if (pivotIndex < offset) {
      select(pivotIndex + 1, right, offset);
    } else {
      select(left, pivotIndex - 1);
    }
  }

  function partition(left, right) {
    const [pivot] = pairs[right];
    let cur = left;
    let leftPartIndex = left;
    console.log(pivot);
    while (cur < right) {
      if (pairs[cur][0] < pivot) {
        swap(leftPartIndex, cur);
      }
    }
    swap(right, leftPartIndex);
    return leftPartIndex;
  }

  function swap(x, y) {
    const term = pairs[x];
    pairs[x] = pairs[y];
    pairs[y] = term;
  }
};
// @lc code=end

topKFrequent([4, 1, -1, 2, -1, 2, 3], 2);
